components/TextFields/src/MDCTextInputController.h

```diff
@@ -0,0 +1,204 @@
+   count / max below text.
+   */
+  MDCTextInputPresentationStyleDefault = 0,
+
+  /**
+   The placeholder text is laid out inline but will float above the field when there is content or
+   the field is being edited. The character count is below text. The Material Design guidelines
+   call this 'Floating inline labels.'
+   https://material.io/guidelines/components/text-fields.html#text-fields-labels
+   */
+  MDCTextInputPresentationStyleFloatingPlaceholder,
+
+  /** The placeholder is laid out inline and the character count is also inline to the right. */
+  MDCTextInputPresentationStyleFullWidth,
+};
+
+/**
+ Material Design compliant states for textInputs. The logic for 'automagic' error states changes:
+ underline color, underline text color.
+ https://www.google.com/design/spec/components/text-fields.html#text-fields-single-line-text-field
+ */
+@interface MDCTextInputController : NSObject <MDCTextInputPositioningDelegate, NSCoding, NSCopying>
+
+/**
+ The character counter. Override to use a custom character counter.
+
+ Default is an internal instance MDCTextInputAllCharactersCounter. Setting this property to null
+ will reset it to return that instance.
+ */
+@property(nonatomic, null_resettable, weak) IBInspectable id<MDCTextInputCharacterCounter>
+    characterCounter;
+
+/**
+ The character count maximum for the text input. A label under the input counts characters entered
+ and presents the count / the max.
+
+ If character count / max has been hidden by the characterCountViewMode
+ (ie: UITextFieldViewModeNever) changing the value of characterLimit has no effect.
+
+ If the character count goes above its max, the underline, the character count / max label and
+ any floating placeholder label all turn to the error color; the text input will be in error state.
+ Note: setErrorText:errorAccessibilityValue: also sets these MDCTextInput properties.
+
+ There is no support for a minimum character count.
+
+ Default is 0.
+ */
+@property(nonatomic, assign) IBInspectable NSUInteger characterCountMax NS_SWIFT_NAME(characterMax);
+
+/**
+ Controls when the character count will be shown and therefore whether character counting determines
+ error state.
+
+ Default is UITextFieldViewModeNever.
+ */
+@property(nonatomic, assign) UITextFieldViewMode characterCountViewMode NS_SWIFT_NAME(characterMode)
+    ;
+
+/**
+ The color used to denote error state in the underline, the errorText's label, the placeholder and
+ the character count label.
+
+ Default is red.
+ */
+@property(nonatomic, nullable, strong) UIColor *errorColor UI_APPEARANCE_SELECTOR;
+
+/**
+ The color applied to the placeholder when floating. However, when in error state, it will be
+ colored with the error color.
+
+ Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, nullable, strong)
+    UIColor *floatingPlaceholderColor NS_SWIFT_NAME(floatingColor) UI_APPEARANCE_SELECTOR;
```

Comment:
Here and above: You're dropping a lot of information in the rename, the "Placeholder" bit seems pretty important. I'd just keep the ObjC name, and if you feel it's too long, you should ideally try to rename in ObjC.